src/App/Fossa/VSI/DynLinked/Internal/Lookup/APK.hs

```diff
@@ -0,0 +1,155 @@
+apkTactic (Just APKLookupTable{..}) file = do
+  index <- Map.lookup file pathToIndex
+  meta <- Map.lookup index indexToMeta
+  Just . DynamicDependency file . Just $ ResolvedLinuxPackage LinuxPackageManagerRPM meta
+apkTactic _ _ = Nothing
+
+-- | The output of the syft binary
+newtype SyftData = SyftData
+  { syftArtifacts :: [SyftArtifact]
+  }
+  deriving (Eq, Show)
+
+instance FromJSON SyftData where
+  parseJSON = withObject "SyftData" $ \obj ->
+    SyftData . filter isAPKArtifact <$> obj .: "artifacts"
+
+data SyftArtifact = SyftArtifact
+  { artifactName :: Text
+  , artifactVersion :: Text
+  , artifactType :: Text
+  , artifactMetadataType :: Text
+  , -- We may run this on systems containing packages whose type we don't handle.
+    -- Delay parsing the metadata until it is required.
+    artifactMetadata :: Value
+  }
+  deriving (Eq, Show)
+
+instance FromJSON SyftArtifact where
+  parseJSON = withObject "SyftArtifact" $ \obj ->
+    SyftArtifact <$> obj .: "name"
+      <*> obj .: "version"
+      <*> obj .: "type"
+      <*> obj .: "metadataType"
+      <*> obj .: "metadata"
+
+data SyftArtifactMetadata = SyftArtifactMetadata
+  { metadataArchitecture :: Text
+  , metadataFiles :: [SyftArtifactMetadataFile]
+  }
+  deriving (Eq, Show)
+
+instance FromJSON SyftArtifactMetadata where
+  parseJSON = withObject "SyftArtifactMetadata" $ \obj ->
+    SyftArtifactMetadata <$> obj .: "architecture"
+      <*> obj .: "files"
+
+instance ToJSON SyftArtifactMetadata where
+  toJSON SyftArtifactMetadata{..} =
+    object
+      [ "architecture" .= metadataArchitecture
+      , "files" .= metadataFiles
+      ]
+
+newtype SyftArtifactMetadataFile = SyftArtifactMetadataFile
+  { metadataFilePath :: Path Abs File
+  }
+  deriving (Eq, Show)
+
+instance FromJSON SyftArtifactMetadataFile where
+  parseJSON = withObject "SyftArtifactMetadataFile" $ \obj ->
+    SyftArtifactMetadataFile <$> obj .: "path"
+
+instance ToJSON SyftArtifactMetadataFile where
+  toJSON SyftArtifactMetadataFile{..} = object ["path" .= metadataFilePath]
+
+isAPKArtifact :: SyftArtifact -> Bool
+isAPKArtifact SyftArtifact{..} = artifactType == "apk" && artifactMetadataType == "ApkMetadata"
+
+runSyft :: (Has Diagnostics sig m, Has Exec sig m) => Path Abs Dir -> m SyftData
+runSyft root = execJson root syftCommand
+
+syftCommand :: Command
+syftCommand =
+  Command
+    { cmdName = "syft"
+    , cmdArgs = ["/", "-o", "json"]
+    , cmdAllowErr = Never
+    }
+
+-- | Multiple file paths can point to the same metadata.
+data APKLookupTable = APKLookupTable
+  { pathToIndex :: Map (Path Abs File) Word
+  , indexToMeta :: Map Word LinuxPackageMetadata
+  }
+  deriving (Eq, Show)
+
+buildLookupTable :: (Has Diagnostics sig m, Has Exec sig m) => Path Abs Dir -> m (Maybe APKLookupTable)
+buildLookupTable root | runningLinux = runMaybeT $ do
+  syft <- MaybeT . recover $ runSyft root
+  compileSyftOutput syft
+buildLookupTable _ = pure Nothing
```

Comment:
I'd like to avoid using monad transformers at all.  This one is unexpected simple to rewrite, though it's not super obvious until you see it (It's bitten me before):
```suggestion
buildLookupTable root | runningLinux = do
  syft <- recover $ runSyft root
  traverse compileSyftOutput syft
buildLookupTable _ = pure Nothing
```

Reply:
Oh! 🧠 yeah, that makes immediate sense. Thanks!